sigstore-java/src/main/java/dev/sigstore/tuf/TufClient.java

```diff
@@ -0,0 +1,284 @@
+      nextVersion++;
+    }
+
+    // 5.3.10) Check expiration timestamp in trusted root is higher than fixed update start time,
+    // otherwise throw error.
+    ZonedDateTime expires = trustedRoot.getSignedMeta().getExpiresAsDate();
+    if (expires.isBefore(updateStartTime)) {
+      throw new RootExpiredException(mirror.toString(), updateStartTime, expires);
+    }
+    // 5.3.11) If the timestamp and / or snapshot keys have been rotated, then delete the trusted
+    // timestamp and snapshot metadata files.
+    if (hasNewKeys(preUpdateSnapshotRole, trustedRoot.getSignedMeta().getRole(Role.Name.SNAPSHOT))
+        || hasNewKeys(
+            preUpdateTimestampRole, trustedRoot.getSignedMeta().getRole(Role.Name.TIMESTAMP))) {
+      File snapshotMetaFile = localStore.resolve("snapshot.json").toFile();
+      if (snapshotMetaFile.exists()) {
+        Files.delete(snapshotMetaFile.toPath());
+      }
+      File timestampMetaFile = localStore.resolve("timestamp.json").toFile();
+      if (timestampMetaFile.exists()) {
+        Files.delete(timestampMetaFile.toPath());
+      }
+    }
+  }
+
+  private boolean hasNewKeys(RootRole oldRole, RootRole newRole) {
+    return newRole.getKeyids().stream().allMatch(s -> oldRole.getKeyids().contains(s));
+  }
+
+  /**
+   * Verifies that a delegate role has been signed by the threshold amount of keys.
+   *
+   * @param signatures these are the signatures on the role meta we're verifying
+   * @param publicKeys a map of key IDs to public keys used for signing various roles
+   * @param role the key ids and threshold values for role signing
+   * @throws SignatureVerificationException if there are not enough verified signatures
+   */
+  @VisibleForTesting
+  static void verifyDelegate(
+      List<Signature> signatures,
+      Map<String, Key> publicKeys,
+      Role role,
+      byte[] verificationMaterial)
+      throws SignatureVerificationException, NoSuchAlgorithmException, InvalidKeyException,
+          InvalidKeySpecException {
+    // use set to not count the same key multiple times towards the threshold.
+    var goodSigs = new HashSet<>(role.getKeyids().size());
+    // role.getKeyIds() defines the keys allowed to sign for this role.
+    for (String keyid : role.getKeyids()) {
+      Optional<Signature> signatureMaybe =
+          signatures.stream().filter(sig -> sig.getKeyId().equals(keyid)).findFirst();
+      // only verify if we find a signature that matcheds an allowed key id.
+      if (signatureMaybe.isPresent()) {
+        var signature = signatureMaybe.get();
+        // look for the public key that matches the key ID and use it for verification.
+        var key = publicKeys.get(signature.getKeyId());
+        if (key != null) {
+          // key bytes are in Hex not Base64!
+          // TODO(patrick): this will change in a subsequent version. Add code to handle PEM Encoded
+          // keys as well.
+          byte[] keyBytes = Hex.decode(key.getKeyVal().get("public"));
```

Comment:
minor comment, and this is style, so feel free to ignore me, i'm only here for logic :)

Why not put the logic for decoding the keybytes into `constructTufPublicKey`? (as in `constructTufPublicKey` takes in the whole `Key`?) that way swapping out the decoding/parsing of the `keyval` can be contained in the `Keys` package, since it will change later.

Reply:
It's a good comment!  I left it that way because I have an eye to refactoring the Keys tool to merge the two Keys public key methods once we have PEM encoded keys here.